src/python/include/Simplex_tree_interface_multi.h

```diff
@@ -0,0 +1,324 @@
+  }
+
+  Skeleton_simplex_iterator get_skeleton_iterator_end(int dimension) {
+	// this specific case works because the range is just a pair of iterators - won't work if range was a vector
+	return Base::skeleton_simplex_range(dimension).end();
+  }
+
+  std::pair<Boundary_simplex_iterator, Boundary_simplex_iterator> get_boundary_iterators(const Simplex& simplex) {
+	auto bd_sh = Base::find(simplex);
+	if (bd_sh == Base::null_simplex())
+	  throw std::runtime_error("simplex not found - cannot find boundaries");
+	// this specific case works because the range is just a pair of iterators - won't work if range was a vector
+	auto boundary_srange = Base::boundary_simplex_range(bd_sh);
+	return std::make_pair(boundary_srange.begin(), boundary_srange.end());
+  }
+
+
+
+// ######################## MULTIPERS STUFF
+  void reset_keys(){
+	unsigned int count = 0;
+	for (auto sh : Base::filtration_simplex_range())
+		Base::assign_key(sh, count++);
+  }
+
+  int get_key(const Simplex& simplex){
+	return Base::key(Base::find(simplex));
+  }
+
+  void set_key(Simplex& simplex, int key){
+	Base::assign_key(Base::find(simplex), key);
+	return;
+  }
+  void fill_lowerstar(std::vector<double> filtration, int axis){
+	for (auto &SimplexHandle : Base::complex_simplex_range()){
+		std::vector<double> current_birth = Base::filtration(SimplexHandle);
+		double to_assign = -1*std::numeric_limits<double>::infinity();
+		for (auto vertex : Base::simplex_vertex_range(SimplexHandle)){
+			to_assign = std::max(filtration[vertex], to_assign);
+		}
+		current_birth[axis] = to_assign;
+		Base::assign_filtration(SimplexHandle, current_birth);
+	}
+  }
+  using simplices_list = std::vector<std::vector<int>>;
+  simplices_list get_simplices_of_dimension(int dimension){
+	simplices_list simplex_list;
+	simplex_list.reserve(Base::num_simplices());
+	for (auto &simplexhandle : Base::skeleton_simplex_range(dimension)){
+		if (Base::dimension(simplexhandle) == dimension){
+			std::vector<int> simplex;
+			simplex.reserve(dimension+1);
+			for (int vertex : Base::simplex_vertex_range(simplexhandle))
+				simplex.push_back(vertex);
+			simplex_list.push_back(simplex);
+		}
+	}
+	simplex_list.shrink_to_fit();
+	return simplex_list;
+  }
+  using edge_list = std::vector<std::pair<std::pair<int,int>, std::pair<double, double>>>;
+  edge_list get_edge_list(){
+	edge_list simplex_list;
+	simplex_list.reserve(Base::num_simplices());
+	for (auto &simplexHandle : Base::skeleton_simplex_range(2)){
+		if (Base::dimension(simplexHandle) == 1){
+			std::pair<int,int> simplex;
+			auto it = Base::simplex_vertex_range(simplexHandle).begin();
+			simplex = {*it, *(++it)};
+			auto f = Base::filtration(simplexHandle);
+			simplex_list.push_back({simplex, {f[0], f[1]}});
+		}
+	}
+	simplex_list.shrink_to_fit();
+	return simplex_list;
+  }
+
+  euler_chars_type euler_char(const std::vector<std::vector<double>> &point_list){ // TODO multi-critical 
+		const unsigned int npts = point_list.size();
```

Comment:
(picking one place randomly)
Using an unsigned type to encode the fact that a number cannot be negative is [not super popular](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es106-dont-try-to-avoid-negative-values-by-using-unsigned) these days, except to match an existing interface. For `npts` it would be natural to use `std::size_t`, what `size()` returns. For `nparameters` which is known to be small, either `size_t` because that's what size returns, or `int` which is the default type for small integers. `unsigned int` tends to be only used when doing bit operations, both because its arithmetic is more error-prone than the corresponding signed type, and because it is often harder to optimize for compilers. Although if you like `unsigned int` we can live with it, we already have quite a few in Gudhi, I just wanted to make sure you were aware of this.

Reply:
OK, I will change that to an int; with all of the other occurrences of unsigned int, as I'm not doing bit operations.